### ISSUE TITLE
Nonsensical "Trying use linear name myLinVar in non-linear context", but not after let expression

### ISSUE THREAD

**Issue by VladimirMarko** [idris-lang/Idris2#1272]
# Steps to Reproduce
Using Idris2 version 0.3.0 I am working through the tutorials. I wrote this:
```idris
module Store

data Access
  = LoggedOut
  | LoggedIn

data Store : Access -> Type where
     MkStore : (secret : String) -> Store st

connect : (1 _ : (1 _ : Store LoggedOut) -> IO a) -> IO a

login : (1 _ : Store LoggedOut) -> (password : String) -> Res Bool (\b => Store (if b then LoggedIn else LoggedOut))

logout : (1 _ : Store LoggedIn) -> Store LoggedOut
readSecret : (1 _ : Store loggedIn) -> Res String (const (Store LoggedIn))

disconnect : (1 _ : Store LoggedOut) -> IO ()

storeProg : IO ()
storeProg = connect $ \s0 => let ok # s1 = login s0 "pwd123" in if ok
  then let secret # s2 = readSecret s1 in
    do
      putStrLn ("Psst: " ++ secret)
      disconnect (logout s2)
  else disconnect s1

```

Idris reports a problem with the `do` block:
```
-UUU:----F1  Store.idr      Bot L24    (Idris2 (Not loaded) Ind ElDoc) -------------------------------------------
- + Errors (1)
 `-- Store.idr line 24 col 25:
     While processing right hand side of storeProg. Trying to
     use linear name s2 in non-linear context.

     Store.idr:24:26--24:28
      20 | storeProg = connect $ \s0 => let ok # s1 = login s0 "pwd123" in if ok
-UUU:%*--F1  *idris2-notes*   Top L1     (Compiler-Notes) --------------------------------------------------------
 20 | storeProg = connect $ \s0 => let ok # s1 = login s0 "pwd123" in if ok
 21 |   then let secret # s2 = readSecret s1 in
 22 |     do
 23 |       putStrLn ("Psst: " ++ secret)
 24 |       disconnect (logout s2)
                               ^^
.
```

However, if I add a `let` line into the middle of the `do` block everything seems fine:
```idris
    do
      putStrLn ("Psst: " ++ secret)
      let whatever = 42
      disconnect (logout s2)
```
Here I get no error message.

If I use `>>` instead of the `do` block:
```idris
    putStrLn ("Psst: " ++ secret) >>
      disconnect (logout s2)
```
I get the same error again.

# Expected Behavior
All 3 variants should compile.

# Observed Behavior
Only the one with the (semantically irrelevant) `let` expression in the middle compiles.

Used source code variants:
[Store.idr - broken do.txt](https://github.com/idris-lang/Idris2/files/6254615/Store.idr.-.broken.do.txt)
[Store.idr - weirdly working do.txt](https://github.com/idris-lang/Idris2/files/6254614/Store.idr.-.weirdly.working.do.txt)
[Store.idr - broken bind.txt](https://github.com/idris-lang/Idris2/files/6254613/Store.idr.-.broken.bind.txt)

**Comment by Russoul** [comment]
Congratulations, you have found a bug that lets us break linearity rules:
```
dupLinear : (1 x : a) -> (a, a)
dupLinear x = dup (let var = 22 in x)
```

I am going to open a new issue.

**Comment by Russoul** [comment]
Regarding your example, that do block should not work in any case when you sequence multiple IO actions.
That's because Monad's `(>>=)` and `(>>)` are not linear.

You can make a small correction to successfully type check your code:
```
storeProg : IO ()
storeProg = connect $ \s0 => let ok # s1 = login s0 "pwd123" in if ok
  then let secret # s2 = readSecret s1 in
    do
      putStrLn ("Psst: " ++ secret) `io_bind` \_ =>
       disconnect (logout s2)
  else disconnect s1

```

i.e. use `io_bind` directly, which is linear.

**Comment by edwinb** [comment]
This behaves as it should now that I've fixed #1274. The error message is about `(>>=)` as @russoul says above. You could also say `let (>>=) = io_bind` to get linear behaviour in the `do` block.
